packages/flutter/lib/src/material/date_picker.dart

```diff
@@ -3129,7 +3174,8 @@ class _InputDateRangePickerState extends State<_InputDateRangePicker> {
   String? _validateDate(DateTime? date) {
     if (date == null) {
       return widget.errorFormatText ?? MaterialLocalizations.of(context).invalidDateFormatLabel;
-    } else if (date.isBefore(widget.firstDate) || date.isAfter(widget.lastDate)) {
+    } else if (date.isBefore(widget.firstDate) || date.isAfter(widget.lastDate) ||
+               widget.selectableDayPredicate != null && !widget.selectableDayPredicate!(date, _startDate, _endDate)) {
```

Comment:
that would be less readable imo; the line should be broken in two 

```
    } else if (date.isBefore(widget.firstDate) || date.isAfter(widget.lastDate) ||
        widget.selectableDayPredicate != null 
        && !widget.selectableDayPredicate!(date, _startDate, _endDate)) {
```
or
```
    } else if (date.isBefore(widget.firstDate) || date.isAfter(widget.lastDate) ||
        widget.selectableDayPredicate != null && 
        !widget.selectableDayPredicate!(date, _startDate, _endDate)) {
```
depending on flutter style guidelines

Reply:
The second one. I don't believe the style guidelines explicitly call it out, but we put conditionals before the line break, not after. I am fine with breaking this into more lines.

Reply:
I'm a bit lost, what should I do ? Just reduce the indentation ? Or also break each condition into different lines ?

Reply:
I personally prefer keeping the null check in the same line as the use of it as they are highly linked

Reply:
The issue is that the second line is split in two, so you might as well split it where it make sens.

```
    } else if (date.isBefore(widget.firstDate) || date.isAfter(widget.lastDate) ||
        widget.selectableDayPredicate != null && 
        !widget.selectableDayPredicate!(date, _startDate, _endDate)) {
```

Reply:
So to make sens, maybe also split the first `||` ?
```dart
    } else if (date.isBefore(widget.firstDate) ||
        date.isAfter(widget.lastDate) ||
        widget.selectableDayPredicate != null && 
        !widget.selectableDayPredicate!(date, _startDate, _endDate)) {
```

This nullability adds so much noise. What do you think if, instead of passing a null `selectableDayPredicate`, we pass a non-nullable predicate with a default to a static private function always returning true ?

Reply:
Like this:
```dart
Future<DateTimeRange?> showDateRangePicker({
    SelectableDayForRangePredicate selectableDayPredicate = _alwaysSelectable,
});
```

Reply:
Or even this to keep a clean api:
```dart
Future<DateTimeRange?> showDateRangePicker({
    SelectableDayForRangePredicate selectableDayPredicate = _alwaysSelectable,
}) {
  // ...
  Widget dialog = DateRangePickerDialog(
    currentDate: currentDate,
    selectableDayPredicate: selectableDayPredicate,
  );
  // ...
}

class DateRangePickerDialog {
  const DateRangePickerDialog({
    SelectableDayForRangePredicate? selectableDayPredicate,
  }) :
      this.selectableDayPredicate = selectableDayPredicate ?? _alwaysSelectable
  ;

  final SelectableDayForRangePredicate selectableDayPredicate;
}
```